preshed/bloom.pyx

```diff
@@ -54,40 +70,124 @@ cdef class BloomFilter:
 
 
 cdef bytes bloom_to_bytes(const BloomStruct* bloom):
-    py = array("L")
-    py.append(bloom.hcount)
-    py.append(bloom.length)
-    py.append(bloom.seed)
-    for i in range(bloom.length // sizeof(key_t)):
-        py.append(bloom.bitfield[i])
-    if hasattr(py, "tobytes"):
-        return py.tobytes()
-    else:
-        # Python 2 :(
-        return py.tostring()
+    cdef key_t pad = 0 # to differentiate new from old data format
+    cdef key_t buflen
+    prefix = struct.pack(FORMAT, pad, VERSION, bloom.hcount, bloom.length, bloom.seed)
+    # note that the modulus check is only required for data that has come from
+    # legacy deserialization - otherwise length is always a multiple of
+    # KEY_BITS.
+    buflen = bloom.length // KEY_BITS
+    if bloom.length % KEY_BITS > 0:
+        buflen += 1
+    buffer = (<char*>bloom.bitfield)[0:buflen * sizeof(key_t)]
+    return prefix + buffer
 
 
 cdef void bloom_from_bytes(Pool mem, BloomStruct* bloom, bytes data):
-    py = array("L")
-    if hasattr(py, "frombytes"):
-        py.frombytes(data)
-    else:
-        py.fromstring(data)
-    bloom.hcount = py[0]
-    bloom.length = py[1]
-    bloom.seed = py[2]
-    bloom.bitfield = <key_t*>mem.alloc(bloom.length // sizeof(key_t), sizeof(key_t))
-    for i in range(bloom.length // sizeof(key_t)):
-        bloom.bitfield[i] = py[3+i]
+    # new-style memory structure (each unit is a key_t/64 bits):
+    # - pad1: 0 (not valid in old data)
+    # - VERSION: 1 (can be raised later if necessary
+    # (following values are same as old-style, except length semantics changed)
+    # - hcount: number of hashes
+    # - length: bitfield length in bits
+    # - seed: seed value for hashes
+
+    if len(data) < STRUCT_SIZE:
+        # unlikely but possible with old data
+        bloom_from_bytes_legacy(mem, bloom, data)
+        return
+    pad, ver, hcount, length, seed = struct.unpack(FORMAT, data[0:STRUCT_SIZE])
+    if pad !=0:
+        bloom_from_bytes_legacy(mem, bloom, data)
+        return
+    assert ver == VERSION, "Unknown serialization version"
```

Comment:
This wouldn't really occur as a programming error, but if we have a future version 2 and someone loads it with an older version. Since this is a user-facing error, it should probably raise a 'normal' exception (probably `ValueError`).